tests/integration-tests/clusters_factory.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file.
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import logging
+
+import boto3
+import configparser
+from retrying import retry
+
+from utils import random_alphanumeric, retry_if_subprocess_error, run_command
+
+
+class Cluster:
+    """Contain all static and dynamic data related to a cluster instance."""
+
+    def __init__(self, name, config_file, ssh_key):
+        self.name = name
+        self.config_file = config_file
+        self.ssh_key = ssh_key
+        self.config = configparser.ConfigParser()
+        self.config.read(config_file)
+        self.__cfn_outputs = None
+
+    @property
+    def cfn_name(self):
+        """Return the name of the CloudFormation stack associated to the cluster."""
+        return "parallelcluster-" + self.name
+
+    @property
+    def region(self):
+        """Return the aws region the cluster is created in."""
+        return self.config.get("aws", "aws_region_name", fallback="us-east-1")
+
+    @property
+    def master_ip(self):
+        """Return the public ip of the cluster master node."""
+        return self.cfn_outputs["MasterPublicIP"]
+
+    @property
+    def os(self):
+        """Return the os used for the cluster."""
+        cluster_template = self.config.get("global", "cluster_template", fallback="default")
+        return self.config.get("cluster {0}".format(cluster_template), "base_os", fallback="alinux")
+
+    @property
+    def cfn_outputs(self):
+        """
+        Return the CloudFormation stack outputs for the cluster.
+        Outputs are retrieved only once and then cached.
+        """
+        if self.__cfn_outputs:
```

Comment:
This can be simplified as:

```python
if not self.__cfn_outputs:
    self.__cfn_outputs = self.__retrieve_cfn_outputs()
return self.__cfn_outputs
```

Reply:
done!